src/GafferScene/Blocker.cpp

```diff
@@ -0,0 +1,199 @@
+//
+//  Copyright (c) 2019, Image Engine Design Inc. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "GafferScene/Blocker.h"
+
+#include "GafferScene/SceneNode.h"
+#include "GafferScene/ShaderPlug.h"
+
+#include "Gaffer/PlugAlgo.h"
+#include "Gaffer/StringPlug.h"
+#include "Gaffer/TransformPlug.h"
+
+#include "IECoreScene/Shader.h"
+#include "IECoreScene/ShaderNetwork.h"
+
+#include "IECore/MessageHandler.h"
+#include "IECore/NullObject.h"
+
+using namespace Gaffer;
+using namespace GafferScene;
+
+static IECore::InternedString g_blockersSetName( "__blockers" );
+
+IE_CORE_DEFINERUNTIMETYPED( Blocker );
+
+size_t Blocker::g_firstPlugIndex = 0;
+
+Blocker::Blocker( const std::string &name, GafferScene::Shader *shader )
+	:	ObjectSource( name, "blocker" )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+
+	if( !shader )
+	{
+		throw IECore::Exception( boost::str( boost::format( "Blocker must be able to retrieve GafferScene::Shader instance" ) ) );
+	}
+
+	addChild( shader );
+	addChild( new Plug( "parameters" ) );
+	addChild( new ShaderPlug( "__shader" ) );
+
+}
+
+Blocker::~Blocker()
+{
+}
+
+void Blocker::loadShader( const std::string &shaderName, bool keepExistingValues )
+{
+	// Update internal shader
```

Comment:
No _good_ reason. I ran into some issues back when I wrote this and have now looked into what was going on instead of just "making it work somehow". 

Here are two commits fixing things that gave me trouble before. Especially the first one as the parameter setup behaviour was so different from what we're doing on the OSL side. 11bb293 and e2cf6d1.

Assuming you are ok with those two fixes to existing code, these make things look more like an OSLLight unless I'm missing something: b5ff15b and 0adf089

Do you see any problems with my more fundamental changes?